addons/binding/org.openhab.binding.xiaomivacuum/src/main/java/org/openhab/binding/xiaomivacuum/discovery/VacuumDiscovery.java

```diff
@@ -0,0 +1,196 @@
+    }
+
+    private synchronized void discover() {
+        TreeSet<String> broadcastAdresses = getBroadcastAddresses();
+        HashMap<String, byte[]> responses = new HashMap<String, byte[]>();
+        for (String broadcastAdress : broadcastAdresses) {
+            responses.putAll(sendDiscoveryRequest(broadcastAdress));
+        }
+        for (Entry<String, byte[]> i : responses.entrySet()) {
+            logger.trace("Discovery responses from : {}:{}", i.getKey(), Utils.getSpacedHex(i.getValue()));
+            Message msg = new Message(i.getValue());
+            logger.debug("Vacuum time stamp: {}, OH time {}, delta {}", msg.getTimestamp(), LocalDateTime.now(),
+                    LocalDateTime.now().compareTo(msg.getTimestamp()));
+            String token = Utils.getHex(msg.getChecksum());
+            String id = Utils.getHex(msg.getDeviceId());
+
+            ThingUID uid = new ThingUID(XiaomiVacuumBindingConstants.THING_TYPE_MIIO, id);
+            // TODO: placeholder entries. Here test for MiIO things to determine as what thingType it needs to be added
+            if (discoveryServiceCallback.getExistingDiscoveryResult(uid) != null) {
+                logger.debug("Thing {} was already in discovery Inbox", uid.toString());
+            }
+            if (discoveryServiceCallback.getExistingThing(uid) != null) {
+                logger.debug("Thing {} already exists", uid.toString());
+            }
+            logger.debug("Discovered Xiaomi Robot Vacuum {} at {}", id, i.getKey());
+            if (token.equals("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF") || token.equals("00000000000000000000000000000000")) {
+                logger.debug(
+                        "No token discovered for device {}. To discover token reset the vacuum & connect to it's wireless network and re-run discovery",
+                        id);
+                thingDiscovered(DiscoveryResultBuilder.create(uid)
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_HOST_IP, i.getKey())
+                        .withRepresentationProperty(id).withLabel("Discovered Xiaomi Mi IO Device").build());
+            } else {
+                logger.debug("Discovered token for device {}: {} ('{}')", id, token, new String(msg.getChecksum()));
+                thingDiscovered(DiscoveryResultBuilder.create(uid)
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_HOST_IP, i.getKey())
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_TOKEN, token).withRepresentationProperty(id)
+                        .withLabel("Discovered Xiaomi Mi IO Device").build());
+            }
+        }
+    }
+
+    /**
+     * @return broadcast addresses for all interfaces
+     */
+    private TreeSet<String> getBroadcastAddresses() {
+        TreeSet<String> broadcastAddresses = new TreeSet<String>();
+        try {
+            broadcastAddresses.add("224.0.0.1");
+            broadcastAddresses.add("192.168.3.109");
```

Comment:
true.. it allows me to do some development while I'm on holidays. As discovery does not work otherwise over a VPN. This will go prior to the final version